Move resources

stephanlachnit opened the pull request:
Separates code and resources by handling the paths in a separate python file.

stephanlachnit commented:
Note to myself regarding the icon for the dev script: https://developer.gnome.org/gtk3/stable/GtkIconTheme.html#gtk-icon-theme-add-resource-path
Also, https://developer.gnome.org/gtk3/stable/GtkAboutDialog.html#gtk-about-dialog-set-logo-icon-name suggests that we only need to set an icon for the application window (which there is none afaik), and the about dialog will follow automatically.

cvfosammmm commented:
I put the icon in resources/icons in latest master. It should be found there and displayed in the about dialog. Perhaps you can delete it now from /data.

stephanlachnit commented:
Thanks!

stephanlachnit commented:
I'm a little bit confused right now.
In `scripts/setzer.dev` I added the a line to include the icon, however it just doesn't seem to work. For example, I check in `setzer/workspace/viewgtk.py` line 61 if the icon name is found (which it is), but right after that it gives me an error that the icon is not found. Can you reproduce this, and do you have any idea what do to get it working?

Icons can also be loaded via a gresource file (see [here](https://developer.gnome.org/ThemedIcons/)), that would simplify everything, so if I don't find a solution I'll just go straight away that path.

cvfosammmm commented:
Not sure I understand. So setzer.dev doesn't print true? For me it does.

You don't have to set an icon for the window, GtkApplication seems to do that automatically after setting an application id (in __main__.py).

Btw. maybe you must rename setzer.dev -> setzer.py

cvfosammmm commented:
Can you add the resources path to the service locator module?

stephanlachnit commented:
The weird thing is that is outputs true even though it doesn't find the icon (at least for me). I'll try to rename it.

> Can you add the resources path to the service locator module instead of a global variable?

Sure.

cvfosammmm commented:
I just tried couple things and I think it might be impossible to extend the search path for app icons unfortunately.

stephanlachnit commented:
Okay, I guess I will go ahead with a gresource file then, that should work with the icons.

cvfosammmm commented:
I don't think it will work, but you can try. The icon path works actually, it's just the app icons seemingly have to be in a different path that can't be extended, if I'm not mistaken. Also gresource files aren't really much use with Python as they are used to bundle things in executable files (binaries) mainly, as I understand it from the docs.

stephanlachnit commented:
I'm also working sometimes on Piper, there we use the gresource file to bundle all svgs, just to simplify the code to load the data (explicit path only once necessary), and building the gresource file is pretty easy in meson. Anyway, I'll try and look if it works.

cvfosammmm commented:
Ok, good luck.

stephanlachnit commented:
So the gresource file didn't work, I guess the Gnome Wiki entry really only works in C. Anyway, I just added a `appicon_path` variable, it doesn't really change anything on an installation (path already in the search paths), but fixes the icon in the about dialog.
Once I moved the two settings to the service locator module I should be ready to merge.

Also, I think `own-quotes.svg` and `own-quotes.symbolic.png` aren't necessary?

cvfosammmm commented:
Yes, I removed the .png

stephanlachnit commented:
Can you help me with the service locator module? It's the last thing missing and I don't really get how it works...
I checked the "Allow edits from maintainers" box, so you should be able to push on my branch, if you want to do that.

cvfosammmm commented:
Can you change the meson requirement to something lower? I have 0.49 for example and it seems to work.

cvfosammmm commented:
I also get:
```meson.build:18: WARNING: Passed invalid keyword argument "modules".
WARNING: This will become a hard error in the future.```

Can you fix it?

cvfosammmm commented:
In the main controller put

```
        # init resources path
        ServiceLocator.init_resources_path('@resourcespath@')
        ServiceLocator.init_app_icons_path('@appiconspath@')
```

In ServiceLocator

```
    def init_resources_path(resources_path):
        ServiceLocator.main_window = main_window
```

and

```
    def get_resources_path():
        return ServiceLocator.resources_path
```

Do the same with app icons. Then import and use the ServiceLocator everywhere like

```
ServiceLocator.get_resources_path()
```

stephanlachnit commented:
Thanks for the explanations, I added everything, however there is a problem:

```python
Traceback (most recent call last):
  File "./scripts/setzer.dev", line 24, in <module>
    import builddir.setzer_dev
  File "./scripts/../builddir/setzer_dev.py", line 28, in <module>
    from setzer.workspace.workspace import Workspace
  File "./scripts/../setzer/workspace/workspace.py", line 22, in <module>
    from setzer.document.document import Document, LaTeXDocument, BibTeXDocument
  File "./scripts/../setzer/document/document.py", line 25, in <module>
    import setzer.document.document_builder as document_builder
  File "./scripts/../setzer/document/document_builder.py", line 20, in <module>
    import setzer.document.build_system.build_system as build_system
  File "./scripts/../setzer/document/build_system/build_system.py", line 30, in <module>
    from setzer.app.service_locator import ServiceLocator
  File "./scripts/../setzer/app/service_locator.py", line 30, in <module>
    import setzer.dialogs.document_wizard.document_wizard as document_wizard
  File "./scripts/../setzer/dialogs/document_wizard/document_wizard.py", line 27, in <module>
    from setzer.dialogs.document_wizard.pages.page_document_class import DocumentClassPage
  File "./scripts/../setzer/dialogs/document_wizard/pages/page_document_class.py", line 24, in <module>
    from setzer.app.service_locator import ServiceLocator
ImportError: cannot import name 'ServiceLocator' from partially initialized module 'setzer.app.service_locator' (most likely due to a circular import) (./scripts/../setzer/app/service_locator.py)
```
So yeah the problem is a circular import, `ServiceLocator` need `document_wizard` to init, which needs `DocumentClassPage` which needs `ServiceLocator`. Since I don't know most of the code, do you think there is a way to get this circle resolved?

> Can you change the meson requirement to something lower? I have 0.49 for example and it seems to work. [...]
> `meson.build:18: WARNING: Passed invalid keyword argument "modules".`
> Can you fix it?

The error only appears for Meson <0.51, but I added a workaround for older meson versions.

cvfosammmm commented:
You could pass the resource path to the dialog on init in the ServiceLocator module. I did the same with settings. Then pass it to the Page when that is initialized. Can you do that? In the longer term it could make sense to separate the dialogs from ServiceLocator though, but I wouldn't do it in this PR.

stephanlachnit commented:
Ah I see, I think it isn't possible easily though with this though (at least I didn't get it to work). It is possible to replace variables in any file (i.e. `service_locator.py`), but I would only do this in the main file but else we can't just copy the setzer dir during installation and need to install each file on their own.
I also noticed that you change the version in `dialogs/about/about.py` every time you bump it.
What about adding a `setzer.app.static` file where all static information is stored, including the version and the paths? Since they aren't changed at all, I don't they need to be implemented that complicated, we can init all the vars directly from the main file and forget about it.

cvfosammmm commented:
Just put these lines in `init_dialogs`:

```
        resources_path = ServiceLocator.get_resources_path()
...
        ServiceLocator.dialogs['document_wizard'] = document_wizard.DocumentWizard(main_window, workspace, settings, resources_path)
```

document_wizard.py:

```
    def __init__(self, main_window, workspace, settings, resources_path):
...
        self.resources_path = resources_path
...
        self.pages.append(DocumentClassPage(self.current_values, resources_path))
```

page_document_class.py:

```
    def __init__(self, current_values, resources_path):
...
        self.resources_path = resources_path
```

cvfosammmm commented:
I want to bundle global state in the ServiceLocator for now. But I will refactor it after this PR, put the dialogs in a separate module and get rid of the variable passing.

cvfosammmm commented:
I think the version number should also be in ServiceLocator and initialized in the main controller, but it must not be in this PR.

stephanlachnit commented:
Still doesn't work for me, same error (pushed your suggested changes). Btw if you add my fork to your remote you can also push on this branch if you got it working.

> I want to bundle global state in the ServiceLocator for now.

Wouldn't that still be possible with that approach? e.g.:
```python
@staticmethod
def get_resources_path():
    return static.get_resources_path()
```

It's a bit cheating though since the information isn't really stored in the ServiceLocator.

> I think the version number should also be in ServiceLocator and initialized in the main controller, but it must not be in this PR.

Once we get this to work it should be just a couple of lines doing the same, I can do a PR after this.

cvfosammmm commented:
It doesn't work because you still have the circular import. You must pass the variable to DocumentClassPageView too, didn't see that on first glance.

cvfosammmm commented:
page_document_class.py:

```
DocumentClassPageView(resources_path)
...
class DocumentClassPageView(PageView):

    def __init__(self, resources_path):
```

Then just use `resources_path` instead of `ServiceLocator.get_resources_path()`.

cvfosammmm commented:
I think you must do similar changes in other "pages", and probably other places too.

stephanlachnit commented:
Ok I think it got it know. I added the version as well, since it was basically the same again. Starts without an error. Only problem: the icon in the about dialog is gone again ...
Since the about dialog is broken for me anyway, can you check it?

stephanlachnit commented:
nvm, it was a typo in a function, everything should work now.

cvfosammmm commented:
I get an error:
```
meson.build:22:59: ERROR:  Unknown variable "py3".
```

cvfosammmm commented:
In README you refer to `setzer.dev`, but the file is now called `setzer_dev.py`.

stephanlachnit commented:
Ah yes I messed that up, should be "py".

stephanlachnit commented:
Fixed and rebased
